File: 1472-design-browser-history/1472-design-browser-history.cpp
// SC- O(n)

class BrowserHistory {
private:
    vector<string> history;
    int curr_page;
public:
    BrowserHistory(string homepage) {
        history.push_back(homepage);
        curr_page=0;
    }
    
    void visit(string url) {
        // TC - O(1) Amortized
        // remove forward history if it exists
        if(curr_page<history.size()-1)
            history.resize(curr_page+1);
        
        history.push_back(url);
        curr_page = history.size()-1;
    }
    
    string back(int steps) {
        // TC - O(1)
        curr_page = max(0, curr_page-steps);
        return history[curr_page];
    }
    
    string forward(int steps) {
        // TC- O(1)
        curr_page = min((int)history.size()-1, curr_page+steps);
        return history[curr_page];
    }
};

/**
 * Your BrowserHistory object will be instantiated and called as such:
 * BrowserHistory* obj = new BrowserHistory(homepage);
 * obj->visit(url);
 * string param_2 = obj->back(steps);
 * string param_3 = obj->forward(steps);
 */

File: 1472-design-browser-history/1472-design-browser-history.py
# SC- O(n)

class BrowserHistory:

    def __init__(self, homepage: str):
        self.history = [homepage]
        self.curr = 0

    def visit(self, url: str) -> None:
        # TC- O(n)
        # remove forward history
        self.history = self.history[:self.curr+1]

        self.history.append(url)
        self.curr += 1

    def back(self, steps: int) -> str:
        # TC- O(1)
        self.curr = max(0, self.curr-steps)
        return self.history[self.curr]

    def forward(self, steps: int) -> str:
        # TC- O(1)
        self.curr = min(len(self.history)-1, self.curr+steps)
        return self.history[self.curr]


# Your BrowserHistory object will be instantiated and called as such:
# obj = BrowserHistory(homepage)
# obj.visit(url)
# param_2 = obj.back(steps)
# param_3 = obj.forward(steps)